File: packages/app-launcher/android/src/main/java/io/capawesome/capacitorjs/plugins/applauncher/AppLauncherPlugin.java
package io.capawesome.capacitorjs.plugins.applauncher;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import io.capawesome.capacitorjs.plugins.applauncher.classes.CustomException;
import io.capawesome.capacitorjs.plugins.applauncher.classes.options.CanOpenUrlOptions;
import io.capawesome.capacitorjs.plugins.applauncher.classes.options.OpenUrlOptions;
import io.capawesome.capacitorjs.plugins.applauncher.classes.results.CanOpenUrlResult;
import io.capawesome.capacitorjs.plugins.applauncher.classes.results.OpenUrlResult;
import io.capawesome.capacitorjs.plugins.applauncher.interfaces.NonEmptyResultCallback;
import io.capawesome.capacitorjs.plugins.applauncher.interfaces.Result;

@CapacitorPlugin(name = "AppLauncher")
public class AppLauncherPlugin extends Plugin {

    public static final String ERROR_UNKNOWN_ERROR = "An unknown error has occurred.";
    public static final String TAG = "AppLauncherPlugin";

    private AppLauncher implementation;

    @Override
    public void load() {
        super.load();
        this.implementation = new AppLauncher(this);
    }

    @PluginMethod
    public void canOpenUrl(PluginCall call) {
        try {
            CanOpenUrlOptions options = new CanOpenUrlOptions(call);
            NonEmptyResultCallback<CanOpenUrlResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull CanOpenUrlResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.canOpenUrl(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    @PluginMethod
    public void openUrl(PluginCall call) {
        try {
            OpenUrlOptions options = new OpenUrlOptions(call);
            NonEmptyResultCallback<OpenUrlResult> callback = new NonEmptyResultCallback<>() {
                @Override
                public void success(@NonNull OpenUrlResult result) {
                    resolveCall(call, result);
                }

                @Override
                public void error(Exception exception) {
                    rejectCall(call, exception);
                }
            };
            implementation.openUrl(options, callback);
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void rejectCall(@NonNull PluginCall call, @NonNull Exception exception) {
        String message = exception.getMessage();
        if (message == null) {
            message = ERROR_UNKNOWN_ERROR;
        }
        String code = null;
        if (exception instanceof CustomException) {
            code = ((CustomException) exception).getCode();
        }
        Logger.error(TAG, message, exception);
        call.reject(message, code);
    }

    private void resolveCall(@NonNull PluginCall call) {
        call.resolve();
    }

    private void resolveCall(@NonNull PluginCall call, @Nullable Result result) {
        if (result == null) {
            call.resolve();
        } else {
            call.resolve(result.toJSObject());
        }
    }
}

File: packages/app-launcher/ios/Plugin/AppLauncherPlugin.swift
import Foundation
import Capacitor

@objc(AppLauncherPlugin)
public class AppLauncherPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AppLauncherPlugin"
    public let jsName = "AppLauncher"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "canOpenUrl", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openUrl", returnType: CAPPluginReturnPromise)
    ]
    public static let tag = "AppLauncherPlugin"

    private var implementation: AppLauncher?

    override public func load() {
        self.implementation = AppLauncher(plugin: self)
    }

    @objc func canOpenUrl(_ call: CAPPluginCall) {
        do {
            let options = try CanOpenUrlOptions(call)
            implementation?.canOpenUrl(options) { result, error in
                if let error = error {
                    self.rejectCall(call, error)
                    return
                }
                self.resolveCall(call, result)
            }
        } catch {
            self.rejectCall(call, error)
        }
    }

    @objc func openUrl(_ call: CAPPluginCall) {
        do {
            let options = try OpenUrlOptions(call)
            implementation?.openUrl(options) { result, error in
                if let error = error {
                    self.rejectCall(call, error)
                    return
                }
                self.resolveCall(call, result)
            }
        } catch {
            self.rejectCall(call, error)
        }
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        CAPLog.print("[", AppLauncherPlugin.tag, "] ", error)
        let code = (error as? CustomError)?.code
        call.reject(error.localizedDescription, code)
    }

    private func resolveCall(_ call: CAPPluginCall, _ result: Result?) {
        if let result = result?.toJSObject() as? JSObject {
            call.resolve(result)
        } else {
            call.resolve()
        }
    }
}
